src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/Designer/ILaunchProfileEditorClientSession.cs

```diff
@@ -0,0 +1,82 @@
+
+using System;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Debug.Designer
+{
+    /// <summary>
+    /// Proffered by the client, for notifications from the server.
+    /// </summary>
+    internal interface ILaunchProfileEditorClientSession
+    {
+        /// <summary>
+        /// <para>
+        /// Informs the client editor that a profile has been added, including all metadata and values.
+        /// </para>
+        /// <para>
+        /// When the client first connects to the server this method will be called multiple times to inform the client of all existing profiles.
+        /// </para>
+        /// </summary>
+        /// <param name="correlationId">
+        /// The correlation ID provided by the client, if this profile is being added as a result of calling <see cref="ILaunchProfileEditorConnection.AddAsync(Guid, string, string)"/>.
+        /// </param>
+        /// <param name="profile">The newly-added <see cref="Profile"/> along with all metadata and properties.</param>
+        /// <returns>
+        /// A Task that completes when the client has received and scheduled the profile addition.
+        /// </returns>
+        Task LaunchProfileAddedAsync(Guid? correlationId, Profile profile);
+
+        /// <summary>
+        /// Informs the client editor that a profile has been removed.
+        /// </summary>
+        /// <param name="correlationId">
+        /// The correlation ID provided by the client, if the profile is being removed as a result of calling <see cref="ILaunchProfileEditorConnection.RemoveAsync(Guid, string)"/>.
+        /// </param>
+        /// <param name="profileName">
+        /// The name of the removed profile.
+        /// </param>
+        /// <returns>
+        /// A Task that completes when the client has received and scheduled the profile removal.
+        /// </returns>
+        Task LaunchProfileRemovedAsync(Guid? correlationId, string profileName);
+
+        /// <summary>
+        /// Informs the client editor that a profile has been renamed.
+        /// </summary>
+        /// <param name="correlationId">
+        /// The correlation ID provided by the client, if the profile is being renamed as a result of calling <see cref="ILaunchProfileEditorConnection.RenameAsync(Guid, string, string)"/>.
+        /// </param>
+        /// <param name="oldProfileName">
+        /// The old profile name.
+        /// </param>
+        /// <param name="newProfileName">
+        /// The new profile name.
+        /// </param>
+        /// <returns>
+        /// A Task that completes when the client has received and scheduled the profile rename.
+        /// </returns>
+        Task LaunchProfileRenamedAsync(Guid? correlationId, string oldProfileName, string newProfileName);
```

Comment:
Given we are using the profile name as an identifier, and we are supporting renaming that identifier, we need strict guarantees about overlapping requests via this interface, and careful handling on both sides of the contract here. For example, the client may attempt to remove a profile using name A, which the server is attempting to rename from A to B. This is unlikely, but I'd prefer a design that didn't allow such a situation in the first place. Using a GUID for each profile, that's never shown to the user and potentially differs across editor sessions would solve this.

Reply:
The server is going to serialize all of these operations, so either A will be removed and the rename will turn into a no-op, or the rename will succeed and the removal will turn into a no-op. I don't see either outcome as being problematic.

Reply:
The problem is that the client is not always in sync with the server, and must relate the identity to the server. I think there's still a race here. Renaming identifiers seems like something to avoid. Is there not another identifier we could use? Even just an opaque numeric one?